### PR TITLE
parametric: Updates the interface for creating span links in parametric apps

### PR DESCRIPTION
**Pull request by mabdinur**
## Motivation

- Updates the interface for creating a span with span links 
- Updates the interface for adding links to spans [after span creation]
- Enables 4/6 span link tests for the python tracer
- Adds tests to validate `v0.4` and `v0.5` span links encoding 
- Decouple SpanLink object from distributing tracing headers.

Note: This PR does not introduce span link support for parametric apps that use GRPC. GRPC apps are broken and will need to be removed/fixed.

Note: This PR does not address issues with generating span links from distributed traces. This will be addressed in a future PR.

## Workflow

1. ⚠️ Create your PR as draft ⚠️
2. Work on you PR until the CI passes (if something not related to your task is failing, you can ignore it)
3. Mark it as ready for review
    * Test logic is modified? -> Get a review from RFC owner. We're working on refining the `codeowners` file quickly.
    * Framework is modified, or non obvious usage of it -> get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

:rocket: Once your PR is reviewed, you can merge it!

🛟 [#apm-shared-testing](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X) 🛟

## Reviewer checklist

* [ ] [Relevant label](https://github.com/DataDog/system-tests/blob/main/docs/CI/labels.md) (`run-parametric-scenario`, `run-profiling-scenario`...) are presents
* [ ] No system-tests internal is modified. Otherwise, I have the approval from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)
* [ ] CI is green, or failing jobs are not related to this change (and you are 100% sure about this statement)
* [ ] A docker base image is modified?
    * [ ] the relevant `build-XXX-image` label is present
    * [ ] To R&P team: locally build and push the image to hub.docker.com 
* [ ] A scenario is added (or removed)?
    * [ ] Get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)
    * [ ] Once merged, add (or remove) it in system-test-dasboard nightly

### CODE REVIEW

File: tests/parametric/test_span_links.py
Comment (PerfectSlayer) [inline tests/parametric/test_span_links.py:46]: This change break the test.
This test was supposed to create a span links between to span instances.
Now, you only test the span link addition. The span link is created within the test, not the client app.

File: tests/parametric/test_span_links.py
Comment (PerfectSlayer) [inline tests/parametric/test_span_links.py:108]: I don' understand this test. The original test ask to create a span links from HTTP headers (using the original way, it has changed to use the headers field of the link object).

Here, you create a first span.
Then a second span from http headers,
Finally, a last span (called second). This one is supposed to have a span link? And used the second created span headers? Why is that?

**Review by robertomonteromiguel — CHANGES_REQUESTED** [review]
are the python tests failures related with these changes?

File: tests/parametric/test_span_links.py
Comment (PerfectSlayer) [inline tests/parametric/test_span_links.py:157]: Same question here: Why the second span is supposed to have link?

File: tests/parametric/test_span_links.py
Comment (PerfectSlayer) [inline tests/parametric/test_span_links.py:206]: `print` command to clean up

File: tests/parametric/test_span_links.py
Comment (PerfectSlayer) [inline tests/parametric/test_span_links.py:233]: I don't think we should do that. We already had this discussion with Bob I think

**Review by PerfectSlayer — CHANGES_REQUESTED** [review]

**Comment by PerfectSlayer** [comment]
> Note: This PR does not introduce span link support for parametric apps that use GRPC. GRPC apps are broken and will need to be removed/fixed.

I only review the test cases here. Waiting for the PR to support gRPC to review the model / logic change.

File: tests/parametric/test_span_links.py
Comment (bwoebi) [inline tests/parametric/test_span_links.py:203]: add_link intentionally has no trace_id. Users of the API are already expected by other functions to have a mapping <span-id, span> in their test suite. The test is meant to create a link from a given span to another. This may be achieved by a dedicated API, by creating a context from the span, by whatever the preferred method in your language is, but the test itself doesn't directly require a capability "create span link with this span id and trace id", but "create span link to this span".

File: tests/parametric/test_span_links.py
Comment (mabdinur) [reply tests/parametric/test_span_links.py:46]: Actually this test still creates a link between two spans in the client app. The main change here is how the link is sent to the client app. Instead of just sending a parent_id and attributes to the client app this test sends all the fields that can be used to create a span link. This makes the test easier to read/maintain.

This test aims to verify that span links are serialized in the expected format. I think that's the key element we need to test.

File: tests/parametric/test_span_links.py
Comment (mabdinur) [reply tests/parametric/test_span_links.py:108]: The naming is confusing. I should have named the spans: "parent", "linked_span", "child". The goal here is to create an integration test for how span links would be set in a normal application. This is the flow I am trying to recreate: 
 - Span 1 is open (root/parent) -> distributed tracing headers containing span 2 are received (first/linked_span) -> Span 3 is created (second/child). Span 3 is a child of Span 1. -> Span 1 and Span 3 are ended

In this flow Span 3 is a child of Span 1 and is LINKED to Span 2. With this flow span links are never set by the test case. It is set in the application. I can refactor this test to make these relationships more clear. I am noticing a bug that is making this test hard to follow. Does this flow make sense? Is this what we should be doing?

File: tests/parametric/test_span_links.py
Comment (mabdinur) [reply tests/parametric/test_span_links.py:157]: I will update this test to make it more clear

File: tests/parametric/test_span_links.py
Comment (mabdinur) [reply tests/parametric/test_span_links.py:203]: Yes this is ideal but we could run into this scenario: 
Span A -> is created, finished, and flushed
Span B -> is created and attempts to link to Span A
  - Span B can not reference span A in a link because it it no longer stored in the client app.

The only way to guarantee this does not happen is to ensure Span A and Span B are both active. In some client apps this means that Span A will always be the parent of Span B. 

In my refactor I was trying to create a scenarios where Span A and Span B can be started/finished independently and the information about the trace_id is not lost. 

Is this a valid concern? If this is not a valid concern, then we can just close this PR

**Comment by mabdinur** [comment]
> > Note: This PR does not introduce span link support for parametric apps that use GRPC. GRPC apps are broken and will need to be removed/fixed.
> 
> I only review the test cases here. Waiting for the PR to support gRPC to review the model / logic change.

I will hold off on adding GRPC support until we can align on whether this change is even necessary. It is likely that we can close this PR and move forward with the current implementation.

File: tests/parametric/test_span_links.py
Comment (mabdinur) [reply tests/parametric/test_span_links.py:233]: Ahhh interesting. So we are no longer flattening attributes in span links. We need to update the RFC: https://docs.google.com/document/d/1RcTjO8ER5FNweJ0rqVcQEbf6tb1PrlPsQ6coZZww8Xg/edit?disco=AAAA_TaDBkU. 

@bwoebi  The functionality was implemented in the python tracer last week. Should we revert this or should we support both approaches?

**Comment by mabdinur** [comment]
> are the python tests failures related with these changes?

Yes. The python tests are expected to pass once 2.5.0 is released (probably later this week)

File: tests/parametric/test_span_links.py
Comment (bwoebi) [reply tests/parametric/test_span_links.py:203]: No it isn't, it's just relevant for the purposes of testing - in a real application, the user would actually hold a reference to a span or spancontext keeping the span(context) in memory even after it's flushed.

File: tests/parametric/test_span_links.py
Comment (bwoebi) [reply tests/parametric/test_span_links.py:233]: I.e. we are taking a piece from the OTel textbook here, which is mandating a json-compatible representation, and in json it's "true" and not "True".

Comment (mabdinur) [reply tests/parametric/test_span_links.py]: I opened this fix for the python tracer: https://github.com/DataDog/dd-trace-py/pull/8123. Once this fix is merged the 4/6 system tests should pass

**Comment by mabdinur** [comment]
Closing this refactor PR. We can revisit this at a later date. This does not block any OKR work. 

Thanks for your reviews @bwoebi and @PerfectSlayer 🙏

**Pull request closed without merging**
